File: 2 - Regression/5.2 - Backword Elimination/backword_elimination_automatic.py
# -*- coding: utf-8 -*-

import numpy as np
import sklearn as sk
import sklearn.preprocessing
import sklearn.model_selection._split
import statsmodels.api
import pandas as pd

# Load data
data = pd.read_csv('C:/DOC/Workspace/Machine Learning A-Z Template Folder/Part 2 - Regression/Section 5 - Multiple Linear Regression/50_Startups - Copy.csv')
x = data.iloc[:, :-1].values
y = data.iloc[:, -1].values

# Categorical data
label_encoder = sk.preprocessing.LabelEncoder()
x[:, 0] = label_encoder.fit_transform(x[:, 0])
one_hot_encoder = sk.preprocessing.OneHotEncoder(categorical_features = [0])
x = one_hot_encoder.fit_transform(x).toarray()

# Dummy variable trap
x = x[:, 1:]

# Feature scaling
scaler = sk.preprocessing.StandardScaler()
x[:, 2:] = scaler.fit_transform(x[:, 2:]) 

# Test - train split
x_train, x_test, y_train, y_test = sk.model_selection._split.train_test_split(x, y, train_size = 0.7)

# Backwords elimination v1
def backwords_elimination(x, y):
    model = statsmodels.api.OLS(exog = x, endog = y).fit()
    r_squared = model.rsquared_adj
    p_values = model.pvalues
    max_p_value = p_values.max()
    max_p_value_index = np.where(p_values == max_p_value)[0][0]
    if max_p_value > 0.05:
        new_x = np.delete(x, max_p_value_index, axis = 1)    
        new_model, selected_x = backwords_elimination(new_x, y)
        new_r_squared = new_model.rsquared_adj
        return (model, x) if r_squared > new_r_squared else (new_model, selected_x)
    else:
        return (model, x)
    
#model, selected_x = backwords_elimination(x_train, y_train)
#model.summary()

# Backwords elimination v2
def backwords_elimination_v2(selected_indices):
    x = x_train[:, selected_indices]
    model = statsmodels.api.OLS(exog = x, endog = y_train).fit()
    r_squared = model.rsquared_adj
    p_values = model.pvalues
    max_p_value = p_values.max()
    max_p_value_index = np.where(p_values == max_p_value)[0][0]
    if max_p_value > 0.05:
        new_indices = selected_indices.copy()
        del new_indices[max_p_value_index]    
        new_model, selected_indices = backwords_elimination_v2(new_indices)
        new_r_squared = new_model.rsquared_adj
        return (model, selected_indices) if r_squared > new_r_squared else (new_model, selected_indices)
    else:
        return (model, selected_indices)
    
model, selected_indices = backwords_elimination_v2(list(range(len(x_train[0]))))

File: 2 - Regression/5.2 - Backword Elimination/backword_elimination_automatic.R
data = read.csv('C:/DOC/Workspace/Machine Learning A-Z Template Folder/Part 2 - Regression/Section 5 - Multiple Linear Regression/50_Startups - Copy.csv')

# Categorical features
state_levels = levels(data$State)
data$State = factor(data$State, levels = state_levels, labels = seq(length(state_levels)))

# Feature scaling
data[2:4] = scale(data[2:4])

# Test - train split
library('caTools')
is_train = sample.split(data$State, SplitRatio = 0.8)
data_train = subset(data, is_train)
data_test = subset(data, !is_train)

# Backword elimination - Not working properly
# The problem is with factors being unpacked in the model but not whem selecting the data
# Wold need to create a one hot encoding as in python
"browser()
selected_indices = seq(length(data_train))
for (i in seq(length(data_train))) {
  selected_data_train = data_train[, selected_indices]
  regressor = lm(formula = Profit ~ ., data = selected_data_train)
  summary = summary(regressor)
  p_values = (summary$coefficients)[, 4]
  adj_r_squared = summary$adj.r.squared
  max_p_value = max(p_values)
  if (max_p_value > 0.05) {
    max_p_value_index = which(p_values == max_p_value)
    new_selected_indeices = selected_indices[-max_p_value_index]
    newModel = lm(formula = Profit ~ ., data = data_train[, new_selected_indeices])
    new_adj_r_squared = summary(newModel)$adj.r.squared
    if (adj_r_squared < new_adj_r_squared) {
      regressor = newModel
      selected_indices = new_selected_indeices
    } 
  } else {  
    break
  }
}
"

